PortScanner/port_scanner.py

```diff
@@ -0,0 +1,325 @@
+import time
+import socket
+import logging
+import multiprocessing
+import configparser
+import socks
+
+from queue import Queue
+from threading import Thread
+from multiprocessing import Process, cpu_count
+
+
+class Worker(Thread):
+    def __init__(self, tasks: Queue):
+        Thread.__init__(self)
+        self.tasks = tasks
+        self.daemon = True
+        self.start()
+
+    def run(self):
+        while True:
+            func, args, kargs = self.tasks.get()
+            try:
+                func(*args, **kargs)
+            except Exception as e:
+                print(e)
+            finally:
+                self.tasks.task_done()
+
+
+class ThreadPool:
+    def __init__(self, num_threads):
+        self.tasks = Queue(num_threads)
+        for _ in range(num_threads):
+            Worker(self.tasks)
+
+    def add_task(self, func, *args, **kargs):
+        self.tasks.put((func, args, kargs))
+
+    def map(self, func, args_list):
+        for args in args_list:
+            self.add_task(func, args)
+
+    def wait_completion(self):
+        self.tasks.join()
+
+
+class Scan:
+
+    """
+
+        Scan class that creates threads and scans a single IP
+        for the given list of ports.
+
+        Arguments:
+
+            target: string - IP Address to scan the ports
+            ports: list - List of ports to scan
+            threads: int - Number of threads to use
+            timeout: int - Socket connection timeout in seconds
+            proxy_ip: list - The proxy IPs to use
+            proxy_port: list - Proxy port to use
+
+        Example:
+
+            scan = Scan(target="192.168.1.0",
+                        ports=range(65536),
+                        threads=200,
+                        timeout=2)
+            scan.run()
+
+    """
+
+    def __init__(self, target: str, ports: list, threads: int=100,
+                 timeout: int=3, proxy_ip: list=["127.0.0.1"],
+                 proxy_port: list=[80]):
+
+        self.target = target
+        self.no_of_threads = threads
+        self.ports = ports
+        self.timeout = timeout
+        self.proxy_ip = proxy_ip
+        self.proxy_port = proxy_port
+
+        self.closed = 0
+        self.opened = 0
+        self.open_ports = []
+        self.closed_ports = []
+
+        self.runtime = 0
+
+    def pscan(self, port: int):
+
+        """
+
+            Port scanning using socket connection. If exception
+            occurs hence the port is closed else it is open.
+
+            Arguments:
+
+                port: int - Port to try connecting to
+
+            Example:
+
+                pscan(3345)
+
+        """
+
+        try:
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.settimeout(self.timeout)
+            con = s.connect((self.target, port))
```

Comment:
Do you feel we can create one sock connection and reuse it for subsequent checks?

Reply:
This was the first approach but it did not work with single sock connection.